fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/SimpleBufferBuilder.java

```diff
@@ -0,0 +1,456 @@
+import static org.lwjgl.opengl.GL32C.*;
+
+/**
+ * A very simple, Mojang inspired BufferBuilder.
+ * <em>This has been customized for 2d rendering such as text and simple planar textures</em>
+ * <p>
+ * Not bound to any specific format, ideally should be held onto for re-use.
+ * <p>
+ * Can be used for 'immediate mode' style rendering using {@link #draw()}, or
+ * upload to external vertex arrays for proper instancing using {@link #finishAndUpload()}.
+ * <p>
+ * This is a Triangles only buffer, all data uploaded is in Triangles.
+ * Quads are converted to triangles using {@code 0, 1, 2, 0, 2, 3}.
+ * <p>
+ * Any given {@link Format} should have its individual {@link Element} components
+ * buffered in the order specified by the {@link Format},
+ * followed by an {@link #endVertex()} call to prepare for the next vertex.
+ * <p>
+ * It is illegal to buffer primitives in any format other than the one specified to
+ * {@link #begin(Format, Mode)}.
+ *
+ * @author covers1624
+ */
+public class SimpleBufferBuilder implements Closeable {
+
+    private static final MemoryUtil.MemoryAllocator ALLOCATOR = MemoryUtil.getAllocator(false);
+
+    private static final int[] VERTEX_ARRAYS = new int[Format.values().length];
+    private static final int[] VERTEX_BUFFERS = new int[Format.values().length];
+    private static final int[] ELEMENT_BUFFERS = new int[Format.values().length];
+
+    static {
+        Arrays.fill(VERTEX_ARRAYS, -1);
+        Arrays.fill(VERTEX_BUFFERS, -1);
+        Arrays.fill(ELEMENT_BUFFERS, -1);
+    }
+
+    private long bufferAddr;   // Pointer to the backing buffer.
+    private ByteBuffer buffer; // ByteBuffer view of the backing buffer.
+
+    private Format format;     // The current format we are buffering.
+    private Mode mode;         // The current mode we are buffering.
+    private boolean building;  // If we are building the buffer.
+    private int elementIndex;  // The current element index we are buffering. if elementIndex == format.types.length, we expect 'endVertex'
+    private int index;         // The current index into the buffer we are writing to.
+    private int vertices;      // The number of complete vertices we have buffered.
+
+    /**
+     * Create a new SimpleBufferBuilder with an initial capacity.
+     * <p>
+     * The buffer will be doubled as required.
+     * <p>
+     * Generally picking a small number, around 128/256 should be a
+     * safe bet. Provided you cache your buffers, it should not mean much overall.
+     *
+     * @param capacity The initial capacity in bytes.
+     */
+    public SimpleBufferBuilder(int capacity) {
+        bufferAddr = ALLOCATOR.malloc(capacity);
+        buffer = MemoryUtil.memByteBuffer(bufferAddr, capacity);
+    }
+
+    public static void destroy() {
+        for (int i = 0; i < VERTEX_ARRAYS.length; i++) {
+            if (VERTEX_ARRAYS[i] == -1) continue; // This set of buffers was never allocated.
+
+            glDeleteBuffers(VERTEX_BUFFERS[i]);
```

Comment:
pass entire array using zeros for unallocated

Reply:
This is NOT working - I tried it and it's just spamming GL errors.